### PR TITLE
rollout support end-to-end canary deployment

### PR DESCRIPTION
**Pull request by zmberg**
### What is end-to-end canary deployment?
The community microservice governance solution MSE, Istio is able to support the end-to-end canary deployment of the business, as follows:
image

### A new end-to-end canary deployment solution
There are some differences between the end-to-end canary deployment model and the regular Rollout model, the most significant of which is that the traffic greyscale should be triggered even if only application C is published. Due to these characteristics, the traffic grayscale should not be tied to a specific application, but should be shared by all applications.

#### The shared TrafficRouting yaml
You can start by defining a **TrafficRouting** resource that describes the rules for full-link greyscale, as follows to import traffic from the Android client into the greyscale environment.

```yaml
apiVersion: rollouts.kruise.io/v1alpha1
kind: TrafficRouting
metadata:
  name: mse-traffic
spec:
  objectRef:
  - service: spring-cloud-a
    ingress:
      name: spring-cloud-a
      classType: mse
  strategy:
    matches:
    - headers:
      - type: RegularExpression
        name: x-user-agent
        value: 'andriod'
```

#### The application's rollouts yaml
For each Application (a, b, c), you can define the corresponding Rollout configuration, and you can add **annotations[rollouts.kruise.io/trafficrouting]** to statement the associated TrafficRouting.

```yaml
---
apiVersion: rollouts.kruise.io/v1alpha1
kind: Rollout
metadata:
  name: rollout-a
  annotations:
    rollouts.kruise.io/trafficrouting: mse-traffic
spec:
  objectRef:
    workloadRef:
      apiVersion: apps/v1
      kind: Deployment
      name: spring-cloud-a
  strategy:
    canary:
      steps:
      - pause: {}
        replicas: 1
---
apiVersion: rollouts.kruise.io/v1alpha1
kind: Rollout
metadata:
  name: rollout-c
  annotations:
    rollouts.kruise.io/trafficrouting: mse-traffic
spec:
  objectRef:
    workloadRef:
      apiVersion: apps/v1
      kind: Deployment
      name: spring-cloud-c
  strategy:
    canary:
      steps:
      - pause: {}
        replicas: 1
---
apiVersion: rollouts.kruise.io/v1alpha1
kind: Rollout
metadata:
  name: rollout-b
  annotations:
    rollouts.kruise.io/trafficrouting: mse-traffic
spec:
  objectRef:
    workloadRef:
      apiVersion: apps/v1
      kind: Deployment
      name: spring-cloud-b
  strategy:
    canary:
      steps:
      - pause: {}
        replicas: 1
```

### CODE REVIEW

File: pkg/trafficrouting/manager.go [hunk 1]
```diff
@@ -52,61 +70,69 @@ func NewTrafficRoutingManager(c client.Client) *Manager {
...
-	// fetch canary service
+	canaryServiceName := getCanaryServiceName(trafficRouting.Service, c.OnlyTrafficRouting)
 	canaryService := &corev1.Service{}
-	err = m.Get(context.TODO(), client.ObjectKey{Namespace: c.Rollout.Namespace, Name: canaryServiceName}, canaryService)
+	canaryService.Namespace = stableService.Namespace
+	canaryService.Name = canaryServiceName
+	// In full-link grayscale scenario(a -> b -> c), if only b or c is released,
+	//and a is not released in this scenario, then the canary service is not needed.
+	if c.OnlyTrafficRouting {
+		goto HandlerNetworkProvider
```
Comment (furykerry) [inline pkg/trafficrouting/manager.go:125]: consider avoid goto, e.g. by extract the service patching logic in a if logic

File: pkg/trafficrouting/network/ingress/ingress.go [hunk 2]
```diff
@@ -214,24 +220,29 @@ func defaultCanaryIngressName(name string) string {
...
-		Weight:        fmt.Sprintf("%d", *weight),
-		Matches:       matches,
-		CanaryService: r.conf.CanaryService,
+		Annotations:           annotations,
+		Weight:                fmt.Sprintf("%d", *weight),
+		Matches:               matches,
+		CanaryService:         r.conf.CanaryService,
+		RequestHeaderModifier: headerModifier,
 	}
+	fmt.Println(util.DumpJSON(data))
```
Comment (furykerry) [inline pkg/trafficrouting/network/ingress/ingress.go:245]: replace Println  with logging

File: pkg/trafficrouting/manager.go [hunk 3]
```diff
@@ -116,78 +142,75 @@ func (m *Manager) DoTrafficRouting(c *util.RolloutContext) (bool, error) {
...
 		trafficRouting.GracePeriodSeconds = defaultGracePeriodSeconds
 	}
 
-	cServiceName := fmt.Sprintf("%s-canary", trafficRouting.Service)
-	trController, err := newNetworkProvider(m.Client, c.Rollout, c.NewStatus, trafficRouting.Service, cServiceName)
+	cServiceName := getCanaryServiceName(trafficRouting.Service, c.OnlyTrafficRouting)
+	trController, err := newNetworkProvider(m.Client, c, trafficRouting.Service, cServiceName)
 	if err != nil {
-		klog.Errorf("rollout(%s/%s) newTrafficRoutingController failed: %s", c.Rollout.Namespace, c.Rollout.Name, err.Error())
+		klog.Errorf("%s newTrafficRoutingController failed: %s", c.Key, err.Error())
```
[the remark this answers is not in the document]
Comment (zmberg) [reply pkg/trafficrouting/manager.go:205]: @sonatype-lift ignoreall

File: pkg/controller/rollout/rollout_progressing.go [hunk 4]
```diff
@@ -149,10 +165,15 @@ func (r *RolloutReconciler) doProgressingInitializing(c *util.RolloutContext) (b
 		klog.Infof("verify rollout(%s/%s) TrafficRouting, and wait a moment", c.Rollout.Namespace, c.Rollout.Name)
 		return false, nil
 	}
+
+	// handler TrafficRouting
```
Comment (furykerry) [inline pkg/controller/rollout/rollout_progressing.go:169]: plz comment about the fact that the traffic-routing here is a gateway that is not directly attach to the rolling application

File: pkg/controller/rollout/rollout_progressing.go [hunk 5]
```diff
@@ -242,13 +263,62 @@ func (r *RolloutReconciler) handleRolloutPlanChanged(c *util.RolloutContext) err
...
 		progressingStateTransition(c.NewStatus, corev1.ConditionTrue, v1alpha1.ProgressingReasonFinalising, "Rollout has been completed and some closing work is being done")
-	} else { // rollout is in rolling
-		return r.canaryManager.runCanary(c)
+		return nil
+	}
+	return r.canaryManager.runCanary(c)
+}
+
+// name is rollout name, tr is trafficRouting name
+func (r *RolloutReconciler) handlerTrafficRouting(namespace, name, tr string) (bool, error) {
```
Comment (furykerry) [inline pkg/controller/rollout/rollout_progressing.go:277]: handlerTrafficRouting -> handleTrafficRouting

File: pkg/controller/rollout/rollout_progressing.go [hunk 6]
```diff
@@ -242,13 +263,62 @@ func (r *RolloutReconciler) handleRolloutPlanChanged(c *util.RolloutContext) err
...
+	err = util.UpdateFinalizer(r.Client, obj, util.AddFinalizerOpType, util.ProgressingRolloutFinalizer(name))
+	if err != nil {
+		klog.Errorf("rollout(%s/%s) add trafficRouting(%s) finalizer failed: %s", namespace, name, tr, err.Error())
+		return false, err
+	}
+	klog.Infof("rollout(%s/%s) add trafficRouting(%s) finalizer(%s) success", namespace, name, tr, util.ProgressingRolloutFinalizer(name))
+	return false, nil
+}
+
+func (r *RolloutReconciler) finalizerTrafficRouting(namespace, name, tr string) error {
```
Comment (furykerry) [inline pkg/controller/rollout/rollout_progressing.go:303]: finalizerTrafficRouting -> finalizeTrafficRouting

File: pkg/trafficrouting/manager.go [hunk 7]
```diff
@@ -52,142 +70,144 @@ func NewTrafficRoutingManager(c client.Client) *Manager {
...
-		klog.Infof("rollout(%s/%s) patch canary service(%s) selector(%s=%s) success",
-			c.Rollout.Namespace, c.Rollout.Name, canaryService.Name, c.Workload.RevisionLabelKey, canaryStatus.PodTemplateHash)
-	}
-	// patch stable service only selector the stable pods
-	if stableService.Spec.Selector[c.Workload.RevisionLabelKey] != canaryStatus.StableRevision {
-		body := fmt.Sprintf(`{"spec":{"selector":{"%s":"%s"}}}`, c.Workload.RevisionLabelKey, canaryStatus.StableRevision)
-		if err = m.Patch(context.TODO(), stableService, client.RawPatch(types.StrategicMergePatchType, []byte(body))); err != nil {
-			klog.Errorf("rollout(%s/%s) patch stable service(%s) selector failed: %s", c.Rollout.Namespace, c.Rollout.Name, stableService.Name, err.Error())
-			return false, err
+		// patch stable service only selector the stable pods
```
Comment (furykerry) [inline pkg/trafficrouting/manager.go:153]: patch stable service only selector the stable pods ->  patch stable service to only select stable pods

File: pkg/trafficrouting/manager.go [hunk 8]
```diff
@@ -52,142 +70,144 @@ func NewTrafficRoutingManager(c client.Client) *Manager {
...
 		}
-	}
 
-	// patch canary service only selector the canary pods
-	if canaryService.Spec.Selector[c.Workload.RevisionLabelKey] != canaryStatus.PodTemplateHash {
-		body := fmt.Sprintf(`{"spec":{"selector":{"%s":"%s"}}}`, c.Workload.RevisionLabelKey, canaryStatus.PodTemplateHash)
-		if err = m.Patch(context.TODO(), canaryService, client.RawPatch(types.StrategicMergePatchType, []byte(body))); err != nil {
-			klog.Errorf("rollout(%s/%s) patch canary service(%s) selector failed: %s", c.Rollout.Namespace, c.Rollout.Name, canaryService.Name, err.Error())
-			return false, err
+		// patch canary service only selector the canary pods
```
Comment (furykerry) [inline pkg/trafficrouting/manager.go:141]: patch canary service only selector the canary pods -> patch canary service to only select canary pods

File: pkg/controller/trafficrouting/trafficrouting_controller.go [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Rollout object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *TrafficRoutingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
Comment (furykerry) [inline pkg/controller/trafficrouting/trafficrouting_controller.go:78]: plz add ut for published func

File: pkg/controller/trafficrouting/trafficrouting_controller.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+	newStatus.ObservedGeneration = tr.Generation
+	return ctrl.Result{}, r.updateTrafficRoutingStatus(tr, *newStatus)
+}
+
+func (r *TrafficRoutingReconciler) updateTrafficRoutingStatus(tr *v1alpha1.TrafficRouting, newStatus v1alpha1.TrafficRoutingStatus) error {
+	if reflect.DeepEqual(tr.Status, newStatus) {
+		return nil
+	}
+	trClone := tr.DeepCopy()
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
Comment (furykerry) [inline pkg/controller/trafficrouting/trafficrouting_controller.go:153]: it is not necessary to retry in reconciliation loop, just re-enqueue is enough

File: pkg/controller/rollout/rollout_progressing.go [hunk 11]
```diff
@@ -149,10 +165,15 @@ func (r *RolloutReconciler) doProgressingInitializing(c *util.RolloutContext) (b
 		klog.Infof("verify rollout(%s/%s) TrafficRouting, and wait a moment", c.Rollout.Namespace, c.Rollout.Name)
 		return false, nil
 	}
+
+	// handler TrafficRouting
+	if c.Rollout.Annotations[v1alpha1.TrafficRoutingAnnotation] != "" {
+		return r.handlerTrafficRouting(c.Rollout.Namespace, c.Rollout.Name, c.Rollout.Annotations[v1alpha1.TrafficRoutingAnnotation])
```
Comment (furykerry) [inline pkg/controller/rollout/rollout_progressing.go:171]: plz add ut in rollout_progressing_test.go to cover rollout with TrafficRoutingAnnotation

File: pkg/controller/trafficrouting/trafficrouting_controller.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+		if strings.Contains(s, v1alpha1.ProgressingRolloutFinalizerPrefix) {
+			progressing.Insert(s)
+		}
+	}
+	return progressing
+}
+
+func (r *TrafficRoutingReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Create a new controller
+	c, err := controller.New("trafficrouting-controller", mgr, controller.Options{
```
[the remark this answers is not in the document]
Comment (zmberg) [reply pkg/controller/trafficrouting/trafficrouting_controller.go:203]: @sonatype-lift ignoreall

File: pkg/trafficrouting/network/ingress/ingress.go [hunk 13]
```diff
@@ -76,47 +78,48 @@ func NewIngressTrafficRouting(client client.Client, conf Config) (network.Networ
 // Initialize verify the existence of the ingress resource and generate the canary ingress
 func (r *ingressController) Initialize(ctx context.Context) error {
 	ingress := &netv1.Ingress{}
-	err := r.Get(ctx, types.NamespacedName{Namespace: r.conf.RolloutNs, Name: r.conf.TrafficConf.Name}, ingress)
```
Comment (furykerry) [inline pkg/trafficrouting/network/ingress/ingress.go:79]: 1. plz update Initialize func to reflect that Initialize only verify the existence of the ingress resource
2. if the ingress resource is generated in EnsureRoutes, will the rollout controller stuck in the Initializing condition?

**Review by furykerry — COMMENTED** [review]
full-link grayscale sounds too chinglish to me, consider replace it with end-to-end canary deployment

File: pkg/trafficrouting/network/ingress/ingress.go [hunk 14]
```diff
@@ -214,23 +217,27 @@ func defaultCanaryIngressName(name string) string {
 	return fmt.Sprintf("%s-canary", name)
 }
 
-func (r *ingressController) executeLuaForCanary(annotations map[string]string, weight *int32, matches []rolloutv1alpha1.HttpRouteMatch) (map[string]string, error) {
+func (r *ingressController) executeLuaForCanary(annotations map[string]string, weight *int32, matches []rolloutv1alpha1.HttpRouteMatch,
+	headerModifier *gatewayv1alpha2.HTTPRequestHeaderFilter) (map[string]string, error) {
```
Comment (furykerry) [inline pkg/trafficrouting/network/ingress/ingress.go:221]: plz add RequestHeaderModifier in later patch

**Pull request merged**